### PR TITLE
refactor(redshift): Improve redshift error handling with new structured reporting system

### PR DESCRIPTION
**Pull request by jjoyce0510**
## Scenarios Covered

Invalid Connection Host Port | ✅
-- | --
Invalid Credentials (user / pass) | ✅
Invalid Permission to Navigate | ✅
Invalid Permission to Extract Profiles | ✅
Invalid Permission to Extract Lineage | ✅
Invalid Permission to Extract Usage | ✅
Failed to Parse Query (Lineage, Usage) | ✅

Also some small fixes

- If the table is NOT external and NOT a view, but has no stats found in svv_table_info during initial scan, then we simply report 0 for row count and size in bytes. This is CORRECT because svv_table_info docs suggest that empty tables will not be present. 
- If the table is EXTERNAL and profile_table_level_only, we can NEVER correctly profile it because EXTERNAL tables are never present in svv_table_info, where we get the "cheap" profiling stats for redshift. We log a warning to the user in such cases to tell them about this behavior. 

## Screenshots
Screenshot 2024-07-08 at 12 10 46 PM
Screenshot 2024-07-08 at 12 21 44 PM
Screenshot 2024-07-08 at 12 21 48 PM
Screenshot 2024-07-08 at 12 23 11 PM
Screenshot 2024-07-08 at 12 49 05 PM
Screenshot 2024-07-08 at 3 59 14 PM
Screenshot 2024-07-08 at 6 25 17 PM

- 
## QA

QA'd locally by attempting to replicate every scenario. Only one I couldn't was lineage parsing.

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

## Summary by CodeRabbit

- **New Features**
  - Improved exception handling and specific error reporting for permission issues and unauthorized access during data profiling.
  - Added detailed error reporting for Redshift lineage extraction failures.
  - Enhanced profiling logic to skip and log messages for external Redshift tables based on certain settings.
  - Added methods for handling Redshift exceptions and reporting failures.

- **Bug Fixes**
  - Addressed issues with profiling tables without necessary attributes by logging warnings and skipping profile creation.

- **Refactor**
  - Updated error reporting formats and enriched table handling in various methods to improve clarity and logging.

- **Chores**
  - Added comments and improved logging for better maintenance and debugging.

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py [hunk 1]
```diff
@@ -1216,8 +1216,22 @@ def _generate_single_profile(
             except Exception as e:
                 if not self.config.catch_exceptions:
                     raise e
-                logger.exception(f"Encountered exception while profiling {pretty_name}")
-                self.report.report_warning(pretty_name, f"Profiling exception {e}")
+
+                error_message = str(e).lower()
+                if "permission denied" in error_message:
+                    self.report.warning(
+                        title="Unauthorized to extract statistics",
```
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:1223]: profiling statistics? i just don't want to confuse people with that

File: metadata-ingestion/src/datahub/ingestion/source/redshift/exception.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        if "svv_table_info" in error_message:
+            report.report_failure(
+                title="Permission denied",
+                message="Failed to extract metadata due to insufficient permission to access 'svv_table_info' table. Please ensure the provided database user has access.",
+                exc=e,
+            )
+        elif "svl_user_info" in error_message:
+            report.report_failure(
+                title="Permission denied",
+                message="Failed to extract metadata due to insufficient permission to access 'svl_user_info' table. Please ensure the provided database user has access.",
```
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/redshift/exception.py:43]: These happen in like 3 places across lineage, usage, and just normal table stuff.

[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/redshift/exception.py:43]: Also the code is a mess in there so I think this fallback is okay.

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:1223]: will accept suggestion

File: metadata-ingestion/src/datahub/ingestion/source/sql/sql_generic_profiler.py [hunk 3]
```diff
@@ -92,16 +92,25 @@ def generate_profile_workunits(
...
-                entityUrn=dataset_urn, aspect=table_level_profile
-            ).as_workunit()
+            if (
+                request.table.column_count is None
+                and request.table.rows_count is None
+                and request.table.size_in_bytes is None
+            ):
+                logger.warning(
+                    f"Table {request.pretty_name} has no column count, rows count, or size in bytes. Skipping emitting table level profile."
+                )
```
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/sql/sql_generic_profiler.py:102]: basically this is skipping emitting a null profile. will check if there are any existing tests i can easily modify

File: metadata-ingestion/src/datahub/ingestion/source/redshift/exception.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/redshift/exception.py:43]: (And this covered the cases I tested with a new sample user who I progressively added privileges for)

[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/redshift/exception.py:43]: You can also see I'm just catching the "redshift" connector errors here as of now, also. So it's fairly targeted to mean that the exception came from a redshift connector call.

**Review by hsheth2 — APPROVED** [review]

**Pull request merged**